near-sdk/src/store/lookup_set/mod.rs

```diff
@@ -105,7 +211,18 @@ where
         T: Borrow<Q>,
         Q: BorshSerialize + ToOwned<Owned = T> + Ord,
     {
-        self.map.remove(value).is_some()
+        let entry = self.get_mut_inner(&value);
+        match entry {
+            EntryState::Inserted => {
+                let _ = std::mem::replace(entry, EntryState::Absent);
```

Comment:
To replace this with absent can actually lead to a bug. Consider the case where the entry is present in storage but not in memory. If you do a put to set the value to inserted in cache, then do this remove, clearing the key at lookup will not happen and the structure will be in inconsistent state. I'm surprised this wasn't caught by the arbitrary tests, but maybe they can be configured better to hit this case.

I tried making a case that will trigger this but it fails for another reason, but maybe it'll help you get on track:
```rust
    #[test]
    fn put_insert_bug() {
        let lookup_key = LookupSet::<u8>::lookup_key(b"m", &8u8, &mut Vec::new());
        {
            // Scoped to make sure set is dropped and persist changes
            let mut set = LookupSet::new(b"m");
            set.put(8u8);
        }

        assert_eq!(crate::env::storage_has_key(&lookup_key), true);
        {
            let mut set = LookupSet::new(b"m");
            set.put(8u8);
            set.remove(&8);
        }
        assert_eq!(crate::env::storage_has_key(&lookup_key), false);
        {
            let set = LookupSet::new(b"m");
            assert!(!set.contains(&8));
        }
    }
```

Reply:
Yes, I came up with this invariant in my head based on the core methods `insert`, `remove` and `contains`. But, as you have noticed, `put`, that I have added later to accommodate for `Extend`, actually breaks it. Thanks for the test and for the thorough review.